support 'make install' for Mono

dsyme opened the pull request:
This supports 'make install' for Mono

As part of our "repo unification" effort https://github.com/Microsoft/visualfsharp/issues/2854  we want to be able to support an installation process that installs the freshly built bits machnie-wide on Linux and OSX. This installation process is also used by typical Linux package build processes (it fakes a root directory and captures which files are installed), and also by homebrew and docker container preparation.

To get things started, the easiest thing to do is to support the existing

    autogen.sh
    make 
    make install

on mono.  I've put as many files as possible under ``mono`` to prevent any pollution of the rest of the build.

Note:  using ``make`` is a bit odd, but it's what's expected by most linux packages, and it's the easiest way to make sure we get exactly the same package preparation as has been used in fsharp/fsharp

Note: it feels like further simplifications can be made to the files being checked in here.  One major round of simplification has already been done and it may be best to just merge then imporve things from here once we have CI in this repo

Note: This PR also adds some files needed to build the "xamarin" versions of FSharp.Core.  We won't yet actually build the final versions of these from this repo (they end up in the FSharp.Core nuget package, which is not yet built from here) though they will be built in CI.  We need to understand better what these variaitions are and if so many of them are really, really needed going forward

dsyme commented:
(Only the Ubuntu build really matters here - the rest should hopefully be unaffected!)

dsyme commented:
Please don't pull until we've discussed with @nosami  and @7sharp9 especially about the files under ``mono\dependencies``.  It would be much better not to add these to this repo. However they were added to http://github.com/fsharp/fsharp and it may be necessary to add them as an interim. But we should really get them from a Mono installation or a nuget package - or we simply avoid building these bespoke FSharp.Core at all

@nosami, @directhex, @7sharp9 could you please check how/what versions of FSharp.Core Xamarin actually needs these days?

nosami commented:
The Android templates are using the NuGet version of FSharp.Core already.

The iOS/WatchOS etc templates are still using the GAC version, but I can't see any reason not to switch them over to using the NuGet version.

This just worked fine for me in the default iOS template. 
```
<packages>
  <package id="FSharp.Core" version="4.0.0.1" targetFramework="xamarinios10" />
</packages>
```
I had [restore errors](https://gist.github.com/nosami/98424dde67fe33aff90693a917baba3a) trying to use 4.1.12.

Are the NuGets built from this repo?

dsyme commented:
@nosami Thanks for the details!

> The Android templates are using the NuGet version of FSharp.Core already.

OK, thanks  But I'm really asking these two questions:

1. why we need a different version (and version number) of FSharp.Core for each of these platforms at all - are they really actually different and could we just use one unified version?  I'll dig around now with ILDASM to see if there are really any substantive differences between these.

2. What is the right way to go about getting the reference assemblies we need to build against?  Currently in http://github.com/fsharp/fsharp we have [checked in a slew of old copies of mscorlib.dll and other DLLs for these different frameworks](https://github.com/fsharp/fsharp/tree/master/mono/dependencies) and we build the various FSharp.Core against those.  It would be better to use a set of proper reference assemblies available on nuget?  I'm trying to avoid checking the binaries into this repo (among other things it bloats the git history)

> The iOS/WatchOS etc templates are still using the GAC version, but I can't see any reason not to switch them over to using the NuGet version. This just worked fine for me in the default iOS template....

OK, that's great, I'd appreciate it if you switched that template over.  Are the templates open source?  We should link to them from the http://github.com/fharp/fsharp README.md

> Are [the FSharp.Core NuGets](https://www.nuget.org/packages/FSharp.Core) built from this repo?

They are currently built using the CI process of http://github.com/fsharp/fsharp using the [spec file here](https://github.com/fsharp/fsharp/tree/master/FSharp.Core.Nuget) but we are very soon going to move that file to this repo and build them as part of CI here 

> I had restore errors trying to use 4.1.12.

We need to understand those better - can you log an issue at http://github.com/fsharp/fsharp please?  They may be related to .NET Standard 1.6 support and represent a bug in nuget or the package.  

FSharp.Core 4.1.12 resolved ok for me (.NET Framework Console app using Visual Studio 2017 on Windows)

```
Attempting to gather dependency information for package 'FSharp.Core.4.1.12' with respect to project 'ConsoleApplication2', targeting '.NETFramework,Version=v4.5.2'
Gathering dependency information took 350.44 ms
Attempting to resolve dependencies for package 'FSharp.Core.4.1.12' with DependencyBehavior 'Lowest'
Resolving dependency information took 0 ms
Resolving actions to install package 'FSharp.Core.4.1.12'
Resolved actions to install package 'FSharp.Core.4.1.12'
Retrieving package 'FSharp.Core 4.1.12' from 'nuget.org'.
Adding package 'FSharp.Core.4.1.12' to folder 'c:\users\dsyme\documents\visual studio 2017\Projects\ConsoleApplication2\packages'
Added package 'FSharp.Core.4.1.12' to folder 'c:\users\dsyme\documents\visual studio 2017\Projects\ConsoleApplication2\packages'
Added package 'FSharp.Core.4.1.12' to 'packages.config'
Successfully installed 'FSharp.Core 4.1.12' to ConsoleApplication2
Executing nuget actions took 2.04 sec
Time Elapsed: 00:00:02.7462500
========== Finished ==========
```

dsyme commented:
@enricosada If possible could you take a look at the problem with resolving FSharp.Core ..12 mentioned above?  It may be something on OSX only though

nosami commented:
> why we need a different version (and version number) of FSharp.Core for each of these platforms at all - are they really actually different and could we just use one unified version? I'll dig around now with ILDASM to see if there are really any substantive differences between these.

Pretty sure that they can be the same version. The FSharp.Core nuget uses the folder `portable-net45+monoandroid10+monotouch10+xamarinios10` for both Android and iOS

> What is the right way to go about getting the reference assemblies we need to build against? Currently in http://github.com/fsharp/fsharp we have checked in a slew of old copies of mscorlib.dll and other DLLs for these different frameworks and we build the various FSharp.Core against those. It would be better to use a set of proper reference assemblies available on nuget? I'm trying to avoid checking the binaries into this repo (among other things it bloats the git history)

This might be the best way at the moment. If you were building on a machine that had Xamarin iOS installed, then you could reference `/Library/Frameworks/Xamarin.iOS.framework/Versions/Current/lib/mono/Xamarin.iOS/mscorlib.dll` etc. These dlls don't exist on nuget right now.

> Are the templates open source? 
I'm afraid not, but I wish they were. Most of the Android / iOS tooling support is proprietary.

dsyme commented:
@nosami What about these? ``xamarinwatchos``, ``xamarintvos``, ``xamarinmacmobile``, ``xamarinmacfull``?  Are they all still a thing?

    <FSCoreVersion Condition="'$(TargetFramework)' == 'xamarinwatchos' ">3.98.41.0</FSCoreVersion>
    <FSCoreVersion Condition="'$(TargetFramework)' == 'xamarintvos' ">3.98.41.0</FSCoreVersion>
    <FSCoreVersion Condition="'$(TargetFramework)' == 'xamarinmacmobile' ">3.99.41.0</FSCoreVersion>
    <FSCoreVersion Condition="'$(TargetFramework)' == 'xamarinmacfull' ">3.100.41.0</FSCoreVersion>

dsyme commented:
@nosami Actually I see these notes:

  <!-- Target MonoTouch: identical to MonoAndroid configurations, though we sanity check by building against the MonoTouch mscorlib -->

  <!-- Target XamarinWatchOS: identical to MonoTouch configurations, though we sanity check by building against the Xamarin.WatchOS mscorlib -->

  <!-- Target XamarinTVOS: identical to MonoTouch configurations, though we sanity check by building against the Xamarin.TVOS mscorlib -->

  <!-- Target xamarinmacmobile similar to monotouch configurations, with Reflection.emit, structural equality, and BigInt -->

  <!-- Target xamarinmacfull should be identical to desktop with missing System.Config, System.Drawing etc -->

dsyme commented:
@nosami Of those, the only ones of interest are ``monoandroid10+monotouch10+xamarinios10`` and ``xamarinmacmobile``.  

Does Xamarin ship F# templates for Xamarin Mac Mobile, or will you want to do that in the future? If so, which ``lib`` folder does it go in in the nuget package? 

The FSharp.Core package doesn't currently include the ``xamarinmacmobile`` build of FSharp.Core - but we could add it.

dsyme commented:
@nosami I have removed the excess variations in https://github.com/Microsoft/visualfsharp/pull/2898/commits/a67d6f085550c1c6f6265812ca38861269a74fa8.  The rules to build ``xamarinmacmobile`` are still there.

dsyme commented:
@KevinRansom assuming this stays all green then this will be ready to merge

dsyme commented on src/fsharp/FSharp.Compiler/FSharp.Compiler.fsproj:
```diff
@@ -532,7 +532,7 @@
     <Reference Include="ISymWrapper, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
     <Reference Include="Microsoft.DiaSymReader.PortablePdb">
-      <HintPath>..\..\..\packages\Microsoft.DiaSymReader.PortablePdb.1.2.0\lib\portable-net45+win8\Microsoft.DiaSymReader.PortablePdb.dll</HintPath>
+      <HintPath>..\..\..\packages\Microsoft.DiaSymReader.PortablePdb.1.1.0\lib\portable-net45+win8\Microsoft.DiaSymReader.PortablePdb.dll</HintPath>
```
@KevinRansom Please check this, I see this in packages.config:
```
  <package id="Microsoft.DiaSymReader.PortablePdb" version="1.1.0"  />
```
This was causing a build failure in fsharp/fsharp

nosami commented:
<img width="622" alt="screen shot 2017-04-26 at 8 53 39 pm" src="https://cloud.githubusercontent.com/assets/667194/25453846/5bf756da-2ac2-11e7-8417-17c035ebce16.png">
I'm not familiar with Mac development myself, so needed to ask

There is a single F# Mac template, but you can target either `full` or `mobile` frameworks with the same template. The mobile version is similar to iOS, small and linkable and can be shipped with the app.

so xamarinmacfull = XM 4.5 (Full) and should have compatibility with "Desktop"

`xamarinmacmobile` should be `xamarinmac20`

dsyme commented:
> xamarinmacmobile should be xamarinmac20

Meaning this entry in the nuspec to put the files in ``lib\xamarinmac20\...`` in the nuget package?

```
    <file src="..\Release\xamarinmac20\bin\FSharp.Core.dll" target="lib\xamarinmac20\FSharp.Core.dll" />
    <file src="..\Release\xamarinmac20\bin\FSharp.Core.xml" target="lib\xamarinmac20\FSharp.Core.xml" />
```

nosami commented:
> Meaning this entry in the nuspec to put the files in lib\xamarinmac20\... in the nuget package?
Yes. Thanks!

dsyme commented:
@nosami Ok we will add that when we next do the nuget package

dsyme commented:
@nosami https://github.com/fsharp/fsharp/pull/733 so we don't forget

7sharp9 commented:
Xamarin.Mac dev could not use a nuget package as far as I can remember as there was no way yo differentiate the version with nuget.

dsyme commented:
> Xamarin.Mac dev could not use a nuget package as far as I can remember as there was no way yo differentiate the version with nuget.

Do you mean Xamarin Mac Full (4.5) or Xamarin Mac Mobile (2.0/2.1)?

dsyme commented:
@KevinRansom I am going to pull this.  Perhaps the most invasive thing about it is the addition of the Makefile (for "make install") and a couple of other files in the root directory.  I hope we can eventually move those to the "mono" directory but I've kept them there for now.

I'd really like to move all stuff specific to .NET Core to a "coreclr" directory too, to keep the root directory much cleaner and allow contributors to mentally ignore what they aren't interested in (e.g. ignore ``mono`` if you're not doing mono, ignore ``coreclr`` if you're not doing .NET Core, ignore ``vs`` if you you're not doing Visual Studio etc.).  I think that's the key to reducing the appearance of complexity.

enricosada commented:
@dsyme please dont move coreclr, i am doing finishing PR to replace that, so just wait a bit.

Also medium term, to remove complexity, i think we are going to use the new fsproj. 
Will work without changes on mono/.net/.netcore using same commands (no more xbuild)

My pr at first will be adding just netcoreapp/netstandard, but we can add multiple framework later because fsproj supprot that already

dsyme commented:
@enricosada No problem, we weren't going to move the files any time soon (though even if we did I suspect that git would handle the move/renaming ok)
